### PR TITLE
Update retained size without causing lock congestion

### PR DESCRIPTION
**Pull request by sopel39**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

Fixes: https://github.com/trinodb/trino/issues/13212

### CODE REVIEW

**Comment by sopel39** [comment]
cc @martint

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java [hunk 1]
```diff
@@ -310,13 +313,40 @@ private boolean addPages(HttpPageBufferClient client, List<Slice> pages)
...
         return true;
     }
 
+    private void updateRetainedMemory()
+    {
+        Lock lock = memoryContextLock.readLock();
+        try {
+            if (memoryContext != null) {
+                // setBytes might be expensive, therefore it needs to be
+                // updated outside of exclusive lock
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java:327]: Maybe move the comment to callspot - this is a bit out of context here

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java [hunk 2]
```diff
@@ -217,12 +223,8 @@ public Slice pollPage()
             return null;
         }
 
-        synchronized (this) {
-            if (!closed) {
-                memoryContext.setBytes(buffer.getRetainedSizeInBytes());
-                scheduleRequestIfNecessary();
-            }
-        }
+        updateRetainedMemory();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java:226]: is that ok to call it if `closed == true`?

**Review by losipiuk — COMMENTED** [review]
LGTM. Thanks

File: core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java (hunk 2 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/DirectExchangeClient.java:228]: > is that ok to call it if closed == true?

It looks like it is

**Comment by sopel39** [comment]
> After this change the contention will likely move to MemoryContext#setBytes.

It will probably move to `io.trino.memory.MemoryPool#reserve`, which is shared across all queries on node. If we have problem with congestion on that level, then we need to fix it holistically, because a lot of operators are updating `MemoryContext`. Maybe `MemoryPool` and `QueryContext` should be lock free.

**Comment by sopel39** [comment]
RN proposal
```
General
* Reduce worker lock congestion. ({issue}`13213`)
```
https://github.com/trinodb/trino/pull/13213

**Comment by colebow** [comment]
Could you go into a little more detail on what lock congestion is? I think we'll need to explain things further in the release notes and tell our users why they should care about this.

**Comment by sopel39** [comment]
> Could you go into a little more detail on what lock congestion is? I think we'll need to explain things further in the release notes and tell our users why they should care about this.

We are still assessing how big of an impact it is. In any case calling `setBytes` within `synchronization` section should have been improved.

**Pull request merged**